### PR TITLE
March work

### PR DESCRIPTION
**Pull request by jeffwarnica**
Class & stdlib conversion: add_vm_to_service, process_telemetry_data (plus, I'm sure, fixes)

core.rb: pulled things from ^

settings/settingsstore: broke apart, allowed for merging settingsstore from multiple domains (projects & local), and documented.

infra_best_fit_with_scope: initial reintegration. Not wired in. Requires testing.

### CODE REVIEW

**Comment by itewk** [comment]
@A-Beck since you are the one most actively still using this can you give it a look?

**Comment by A-Beck** [comment]
@jeffwarnica and @itewk Apologies for the delay. I should be able to test/review this in the coming week.

File: Automate/RedHatConsulting_Utilities/StdLib/Settings.class/__methods__/settings.rb [hunk 1]
```diff
@@ -107,26 +56,124 @@ def get_setting(region, key, default = nil)
...
-            return SETTINGS[region][key] if SETTINGS[region].key?(key)
-            raise(KeyError, "key [#{key}] does not exist in region [#{region}] or defaults settings hash, and no default provided") unless SETTINGS[:default].key?(key)
-            return SETTINGS[:default][key]
-          rescue KeyError => e
-            if default.nil?
-              raise e
-            else
-              return default
+            unless @real_settings.key?(region)
+              raise(KeyError, "region [#{region}] does not exist in settings hash and no default provided for [#{key}]") unless @real_settings[:default].key?(key)
```
Comment (A-Beck) [inline Automate/RedHatConsulting_Utilities/StdLib/Settings.class/__methods__/settings.rb]: Pulled this in to my appliance, and I'm getting a `undefined method 'key?' for nil:NilClass` in vmware_customize:
- https://github.com/RedHatOfficial/miq-Utilities/blob/master/Automate/RedHatConsulting_Utilities/Infrastructure/VM/Provisioning/Placement.class/__methods__/vmware_customize.rb#L52

This seems to be caused by the removal of the default values in `SETTINGS`. Should we keep those older default/global settings for backwards compatibility for those not using the settings feature?

File: Automate/RedHatConsulting_Utilities/Infrastructure/VM/Provisioning/StateMachines/Methods.class/__methods__/add_vm_to_service.rb [hunk 2]
```diff
@@ -2,49 +2,60 @@
...
-#     miq_provision - VM Provisining request contianing the VM to add to a service
+#     miq_provision - VM Provisioning request containing the VM to add to a service
 #
-@DEBUG = false
+module RedHatConsulting_Utilities
+  module Automate
+    module Infrastructure
+      module VM
+        module Provisioning
+          class AddDVMToService
```
Comment (A-Beck) [inline Automate/RedHatConsulting_Utilities/Infrastructure/VM/Provisioning/StateMachines/Methods.class/__methods__/add_vm_to_service.rb]: This class name looks like a typo - should this be changed to `AddVMToService`?

**Comment by A-Beck** [comment]
@jeffwarnica Still having issues with the settings changes when testing in my env ( still on CFME 5.9 FWIW ). 

The only key that seems to be set in the settings hash is `:settings_from`.

Adding logging to https://github.com/jeffwarnica/miq-Utilities/blob/mar21/Automate/RedHatConsulting_Utilities/StdLib/Settings.class/__methods__/settings.rb#L80, it does not seem like the SettingStorage is getting picked up as a descendant for some reason, which would explain why no settings were getting picked up by default. 

Should I need to do anything to have the SettingsStorage class picked up as a descendant?

**Comment by jeffwarnica** [comment]
@A-Beck it is possible 5.9 doesn't support nested embedding? 

Do you recall if RedHatConsulting_Utilities::StdLib::Core::SettingsStorage was not seen at all?

I'll spin up a new lab and see if I can't reproduce in 5.10, or find a 5.9 box somewhere.

**Comment by A-Beck** [comment]
@jeffwarnica I believe `RedHatConsulting_Utilities::StdLib::Core::SettingsStorage ` did not show up at all. I'll do additional testing in my infra to confirm and test out nested embeds on 5.9.

**Comment by A-Beck** [comment]
@jeffwarnica Sorry for the delay here, again.

Looks like you were right, 5.9 has different embedded behavior then in 5.10
In my testing, both the Settings and SettingStorage needed to be imported, in that order, in 5.9.
In 5.10, only Settings needs to be imported, and things work as expected.

Since 5.9 is still supported for another year or so, I don't want to merge this in and break people ( like me ) who may still need to support 5.9 appliances. I also do not see an easy way these settings enhancements can be altered to work across versions.

**Comment by itewk** [comment]
Make a 5.10 branch and merge onto there? And/or also a 5.9 branch? That’s how OpenShift handles these sorts of problems.

**Comment by A-Beck** [comment]
@itewk agreed - that would be the way to do it. I'd be happy to start setting it up.

Have release branches based on CF versions: i.e `release/cfme-v5.9` as the format, perhaps?
Have tags based on CF versions too: `v5.9-1.0.0`, perhaps?

Thoughts?

**Comment by jeffwarnica** [comment]
I'd like to see this pushed, and agree that perhaps the project needs version based streams.  

I have no strong feelings about what that looks like, and can refactor to whatever is decided on.

**Comment by itewk** [comment]
@A-Beck sounds good to me, and you could make master also be the latest,  thats what OCP does anyway. They have the advantage of having their tags line up with their release versions for OCP,  but maybe tricky here, i like your solution. Make it so.

**Comment by A-Beck** [comment]
@itewk good input.
I formalized this in this PR: https://github.com/RedHatOfficial/miq-Utilities/pull/155

Once that is merged in, I'll cut a new major version, and branch 4.6 stuff off that.

@jeffwarnica once https://github.com/RedHatOfficial/miq-Utilities/pull/155 is merged, rebase on master, and we'll get these changes in.

**Comment by A-Beck** [comment]
@jeffwarnica  #155 was merged in. I'll approve this in once the conflict is resolved.

**Comment by A-Beck** [comment]
@jeffwarnica it looks like conflicts are resolved, correct?

I feel like we should merge this in. One more question though - why delete the README linking to README.md? There are minor differences:

```
15:51:23 $ diff README README.md 
12a13
> * [Branches and Tags](#branches-and-tags)
145a147,155
> # Branches and Tags
> The master branch of this repoistory will aim to support the current CloudForms release.
> If breaking changes for previous CloudForms versions are introduced in Master, version specific branches and tags will be created for those previous versions, as long as those releases are still supported by Red Hat. The version-specific release branches will be no longer be maintained once that version of CloudForms is end of life.
> 
> CloudForms Product Lifecycle Information:
> * https://access.redhat.com/support/policy/updates/cloudforms
> 
> # Contributors
> https://github.com/RedHatOfficial/miq-Utilities/graphs/contributors
```

**Comment by jeffwarnica** [comment]
@A-Beck Looks like the conflicts are resolved (and link returned). I've not actually tested this branch ( https://github.com/jeffwarnica/miq-Utilities/tree/fixes_for_quickstart was my working space for the TaT ), but if we get this merged, I can spend some time rebasing and smoke testing other things.

**Comment by A-Beck** [comment]
@jeffwarnica OK, sounds good. I want to get this in - I'll approve and then you can merge it in.

I'll pull this in and start smoketesting with it as well.

**Pull request merged**